viewing_party/party.py

```diff
@@ -1,23 +1,112 @@
+        else None
+    )
+
+
+def add_to_watched(user_data, movie):
+    user_data["watched"].append(movie)
+    return user_data
+
+
+def add_to_watchlist(user_data, movie):
+    user_data["watchlist"].append(movie)
+    return user_data
+
+
+def get_movie_from_watchlist(user_data, movie_title):
+    return next((m for m in user_data["watchlist"] if m["title"] == movie_title), None)
+
+
+def watch_movie(user_data, movie_title):
+    movie = get_movie_from_watchlist(user_data, movie_title)
+    if movie:
+        user_data["watchlist"].remove(movie)
+        return add_to_watched(user_data, movie)
+    else:
+        return user_data
+
 
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
 
 
+def get_watched_avg_rating(user_data):
+    ratings = [movie["rating"] for movie in user_data["watched"]]
+    average = sum(ratings) / len(ratings) if ratings else 0
+    return average
+
+
+def get_most_watched_genre(user_data):
+    genres = [movie["genre"] for movie in user_data["watched"]]
+    most_watched = max(set(genres), key=genres.count, default=None)
+    return most_watched
+
+
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
 
-        
+
+def get_unique_watched(user_data):
+    friends_watched_titles = [
+        movie["title"] for friend in user_data["friends"] for movie in friend["watched"]
+    ]
+    unique_watched = [
+        movie
+        for movie in user_data["watched"]
+        if movie["title"] not in friends_watched_titles
+    ]
+    return unique_watched
+
+
+def get_friends_unique_watched(user_data):
+    watched_titles = [movie["title"] for movie in user_data["watched"]]
+    friends_watched_movies = {
+        movie["title"]: movie
+        for friend in user_data["friends"]
+        for movie in friend["watched"]
+        if movie["title"] not in watched_titles
+    }
+    return friends_watched_movies.values()
+
+
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
 
+
+def get_available_recs(user_data):
+    return [
+        movie
+        for movie in get_friends_unique_watched(user_data)
+        if movie["host"] in user_data["subscriptions"]
+    ]
+
+
 # -----------------------------------------
 # ------------- WAVE 5 --------------------
 # -----------------------------------------
 
+
+def get_new_rec_by_genre(user_data):
+    new_rec_by_genre = [
+        movie
+        for movie in get_friends_unique_watched(user_data)
+        if movie["genre"] is get_most_watched_genre(user_data)
+    ]
+    return new_rec_by_genre
+
+
+def get_rec_from_favorites(user_data):
+    favorite_titles = [movie for movie in user_data["favorites"]]
+    rec_from_favorites = [
+        movie for movie in get_unique_watched(user_data) if movie in favorite_titles

```

Comment:
Though I don't think it's a reasonable assumption, it is definitely possible that there are movies in `user_data["favorites"]` that are not in `user_data["watched"]` that would otherwise fit the criteria for this function, and thus could be missed by calling `get_unique_watched` here.

But as I said, I think it's quite reasonable to assume that a movie is not in `user_data["favorites"]` unless it is also in `user_data["watched"]`. How can your favorite movie be one you haven't watched? (OK, maybe you watched it on some other platform, but still...)